### PR TITLE
allow specialFolder(documents) on linux

### PR DESCRIPTION
**Pull request by mwieder**
Allow specialFolder("documents") to work on linux systems as a step towards achieving parity among platforms.

### CODE REVIEW

**Comment by peter-b** [comment]
Good idea! 👍

At some point it would be nice if the `specialFolder` infrastructure followed the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

**Comment by mwieder** [comment]
Thanks @peter-b - I'm missing having you still on the team.
I'm not sure how much of the XDG variables would be useful in LC and secondarily I'm not sure how many of these can be cross-platform. I had considered mapping "preferences" on linux to .config + stack name, but stopped short of that, at least for now.

File: engine/src/dsklnx.cpp [hunk 1]
```diff
@@ -995,6 +995,12 @@ class MCLinuxDesktop: public MCSystemInterface
             return MCStringFormat(r_folder, "%@/Desktop", *t_home);
         else if (MCNameIsEqualToCaseless(p_type, MCN_home))
             return MCStringCopy(*t_home, r_folder);
+        // 2019-12-19 MDW [[ feature_linux_documents_folder ]] allow linux to have "documents" == ~/Documents
```
Comment (runrevmark) [inline engine/src/dsklnx.cpp]: Current practice is to not put comments like this in the code - as this information should be present in the commit message.

File: engine/src/dsklnx.cpp [hunk 2]
```diff
@@ -995,6 +995,12 @@ class MCLinuxDesktop: public MCSystemInterface
...
+        else if (MCNameIsEqualToCaseless(p_type, MCN_documents))
+            return MCStringFormat(r_folder, "%@/Documents", *t_home);
+        // 2019-12-19 MDW [[ feature_linux_documents_folder ]] allow linux to have "system" == /opt
```
Comment (runrevmark) [inline engine/src/dsklnx.cpp]: I agree with adding support for documents like this - it is no more or less correct than the way we compute the desktop folder after all.

I don't agree with `/opt` being `system` though - given the filesystem hierarchy definition of 'opt' is for "the installation of add-on application software packages", it doesn't seem like a good fit. (It is also out of scope of 'adding support for specialFolderPath(documents)!'.

**Review by runrevmark — CHANGES_REQUESTED** [review]
Hi Mark,

Thanks for this.

For this to be mergeable the following actions need to be done:
- the feedback mentioned in the body of the review should be incorporated
- an enhancement needs to be filed in BZ
- the documentation page for specialFolderPath needs to be updated appropriately
- a release note (tied to the BZ issue number) needs to be added
- since the changes are small, this should be presented as a single commit with suitable title and message

For the commit title/message I suggest something like:
```
[[ Bug NNNN ]] Implement specialFolderPath(documents) on Linux

This patch adds support for the `documents` keyword to the `specialFolderPath`
function using the same (simple) implementation approach as the `desktop`
keyword, i.e. the users home folder is appended with `/Documents`.
```

File: engine/src/dsklnx.cpp (hunk 2 above)
Comment (mwieder) [reply engine/src/dsklnx.cpp]: Looking at this, I agree about /opt. What do you think of the suggestion in https://quality.livecode.com/show_bug.cgi?id=21433 to use /sbin? That seems more reasonable to me.

File: engine/src/dsklnx.cpp (hunk 1 above)
Comment (mwieder) [reply engine/src/dsklnx.cpp]: For the BZ enhancement, does https://quality.livecode.com/show_bug.cgi?id=15094 fit the bill? And is it worth following up on https://quality.livecode.com/show_bug.cgi?id=19236 at this point to account for xdg-user-dirs renaming and localization?

Comment (mwieder) [reply engine/src/dsklnx.cpp]: ...and I'm not going to argue with "current practice", but I do want to make the point that comments in the code avoid the need of having to go look up the commit message to see why this code exists. My two cents worth.

Comment (runrevmark) [reply engine/src/dsklnx.cpp]: Heh - well I did say 'comments like this' - which is more about putting a date / initial / tag and replication of commit message title (metadata at best).

I would never object to good comments which help divide and describe the actual code but there's no point ladening them with metadata one can see elsewhere.

In this case you could do:
```
else if (MCNameIsEqualToCaseless(p_type, MCN_documents))
{
    /* Assume the documents folder is ~/Documents (which is at least true
     * for English localizations). */
    return MCStringFormat(r_folder, "$HOME/Documents", *t_home);
}
```

File: engine/src/dsklnx.cpp (hunk 2 above)
Comment (runrevmark) [reply engine/src/dsklnx.cpp]: Heh - no - Richard gives no use-case nor any indication why /sbin should be used (isn't /sbin in $PATH?).

I can't see how /sbin is anything like /System on macOS, nor is it really much like the Windows folder (as most 'system binaries' I think are actually in a subfolder?). 

Its a separate issue anyway - if someone can come up with a convincing argument as to *why* that specialFolderPath token is useful (on Linux) and by doing so demonstrate that /sbin is the thing to use then I'll happily entertain a patch :)

File: engine/src/dsklnx.cpp (hunk 1 above)
Comment (runrevmark) [reply engine/src/dsklnx.cpp]: Re 15094 - yes that seems to fit the bill exactly :)

I'd suggest going ahead with this patch using the simple approach - then looking at the xdg stuff separately...

This is precisely the suggestion made on this PR https://github.com/livecode/livecode/pull/5164 - as using the C library (gnome?) wrappers around the XDG stuff requires a separate fix elsewhere. (Also that nicely separates the 'adding the feature to specialFolderPath' from 'updating specialFolderPath to work in all localizations on linux' which are logically distinct things).

Comment (mwieder) [reply engine/src/dsklnx.cpp]: > Heh - well I did say 'comments like this' - which is more about putting a date / initial / tag and replication of commit message title (metadata at best).

Ah. My bad then. I thought that (// <date> <initial> [[ tag ]]) was the required form for patch comments.

**Comment by mwieder** [comment]
resubmitted as https://github.com/livecode/livecode/pull/7437, although with some interesting rebase artifacts that I can't seem to shake.

**Comment by mwieder** [comment]
Closing in favor of #7437

**Pull request closed without merging**